### PR TITLE
Refactored temperament tuning

### PR DESCRIPTION
**Pull request by larspalo**
According to discussion in #1333 this is changed to just re-factoring tuning handling code in preparation for a better solution of the issue.

### CODE REVIEW

File: src/core/temperaments/GOTemperament.cpp [hunk 1]
```diff
@@ -13,24 +13,20 @@ GOTemperament::GOTemperament(wxString name, wxString group)
   : m_Group(group),
     m_GroupTitle(wxEmptyString),
     m_Name(name),
-    m_Title(wxEmptyString) {}
+    m_Title(wxEmptyString),
+    m_IsTemperamentOriginalBased(true) {}
```
Comment (oleg68) [inline src/core/temperaments/GOTemperament.cpp:17]: It is better to introduce the third parameter `isOriginalBased = true` to the constructor.
So
```suggestion
    m_IsTemperamentOriginalBased(isOriginalBased) {}
```

File: src/core/temperaments/GOTemperament.cpp [hunk 2]
```diff
@@ -13,24 +13,20 @@ GOTemperament::GOTemperament(wxString name, wxString group)
...
 
 GOTemperament::GOTemperament(
   wxString name, wxString title, wxString group, wxString groupTitle)
-  : m_Group(group), m_GroupTitle(groupTitle), m_Name(name), m_Title(title) {}
+  : m_Group(group),
+    m_GroupTitle(groupTitle),
+    m_Name(name),
+    m_Title(title),
+    m_IsTemperamentOriginalBased(true) {}
```
Comment (oleg68) [inline src/core/temperaments/GOTemperament.cpp:25]:
```suggestion
    m_IsTemperamentOriginalBased(isOriginalBased) {}
```

File: src/core/temperaments/GOTemperament.cpp [hunk 3]
```diff
@@ -43,3 +39,7 @@ wxString GOTemperament::GetGroup() const { return m_Group; }
 wxString GOTemperament::GetGroupTitle() const {
   return m_GroupTitle.IsEmpty() ? wxGetTranslation(m_Group) : m_GroupTitle;
 }
+
+bool GOTemperament::GetIsTemperamentOriginalBased() const {
```
Comment (oleg68) [inline src/core/temperaments/GOTemperament.cpp:43]: This method is good for inlining in the header

File: src/core/temperaments/GOTemperament.h [hunk 4]
```diff
@@ -16,25 +16,20 @@ class GOTemperament {
   wxString m_GroupTitle;
   wxString m_Name;
   wxString m_Title;
+  bool m_IsTemperamentOriginalBased; // To know how tuning should be handled
```
Comment (oleg68) [inline src/core/temperaments/GOTemperament.h:19]:
```suggestion
private:
  bool m_IsTemperamentOriginalBased; // To know how tuning should be handled
```

File: src/core/temperaments/GOTemperament.h [hunk 5]
```diff
@@ -16,25 +16,20 @@ class GOTemperament {
...
 
-  virtual float GetOffset(
-    bool ignorepitch,
-    unsigned midi_number,
-    unsigned wav_midi_number,
-    float wav_pitch_fract,
-    float harmonic_number,
-    float pitch_correction,
-    float default_tuning) const;
+  virtual float GetOffset(unsigned index) const;
```
Comment (oleg68) [inline src/core/temperaments/GOTemperament.h:27]:
```suggestion
  // if m_IsTemperamentOriginalBased then it returns the deviation from the original temperament
  // else it returns the deviation from the equal temperament
  virtual float GetOffset(unsigned index) const;
```

File: src/core/temperaments/GOTemperament.h [hunk 6]
```diff
@@ -16,25 +16,20 @@ class GOTemperament {
...
   wxString GetName() const;
   wxString GetTitle() const;
   wxString GetGroup() const;
   wxString GetGroupTitle() const;
+  bool GetIsTemperamentOriginalBased() const;
```
Comment (oleg68) [inline src/core/temperaments/GOTemperament.h:32]:
```suggestion
  bool IsTemperamentOriginalBased() const { return m_IsTemperamentOriginalBased; }
```

1. The `Is` prefix is sufficient for boolean getters;
2. The method is quite simple for inlining.

File: src/core/temperaments/GOTemperamentCent.cpp [hunk 7]
```diff
@@ -24,6 +24,7 @@ GOTemperamentCent::GOTemperamentCent(
   m_Tuning[9] = 0;
   m_Tuning[10] = 0;
   m_Tuning[11] = 0;
+  m_IsTemperamentOriginalBased = false;
```
Comment (oleg68) [inline src/core/temperaments/GOTemperamentCent.cpp:27]: It's better to pass `false` to the `GOTemperament` constructor.

File: src/core/temperaments/GOTemperamentCent.cpp [hunk 8]
```diff
@@ -53,6 +54,7 @@ GOTemperamentCent::GOTemperamentCent(
   m_Tuning[9] = i10;
   m_Tuning[10] = i11;
   m_Tuning[11] = i12;
+  m_IsTemperamentOriginalBased = false;
```
Comment (oleg68) [inline src/core/temperaments/GOTemperamentCent.cpp:57]: It's better to pass false to the GOTemperament constructor.

File: src/core/temperaments/GOTemperamentCent.cpp [hunk 9]
```diff
@@ -83,23 +85,9 @@ GOTemperamentCent::GOTemperamentCent(
...
```
Comment (oleg68) [inline src/core/temperaments/GOTemperamentCent.cpp:88]: It's better to pass false to the GOTemperament constructor.

File: src/core/temperaments/GOTemperamentCent.h [hunk 10]
```diff
@@ -47,14 +47,7 @@ class GOTemperamentCent : public GOTemperament {
...
 
-  virtual float GetOffset(
-    bool ignorepitch,
-    unsigned midi_number,
-    unsigned wav_midi_number,
-    float wav_pitch_fract,
-    float harmonic_number,
-    float pitch_correction,
-    float default_tuning) const;
+  virtual float GetOffset(unsigned index) const;
```
Comment (oleg68) [inline src/core/temperaments/GOTemperamentCent.h:50]:
```suggestion
  virtual float GetOffset(unsigned index) const override { return m_Tuning[index]; }
```

File: src/core/temperaments/GOTemperamentCent.cpp [hunk 11]
```diff
@@ -83,23 +85,9 @@ GOTemperamentCent::GOTemperamentCent(
...
-  double concert_pitch_correction = 0;
-
-  if (!ignorePitch && wav_midi_number) {
-    concert_pitch_correction = (100.0 * wav_midi_number - 100.0 * midi_number
-                                + log(8.0 / harmonic_number) / log(2) * 1200)
-      + wav_pitch_fract;
-  }
-  return m_Tuning[midi_number % 12] - default_tuning - concert_pitch_correction
-    + pitch_correction;
+float GOTemperamentCent::GetOffset(unsigned index) const {
```
Comment (oleg68) [inline src/core/temperaments/GOTemperamentCent.cpp:91]: It is too simple for placing in the implementation.

File: src/grandorgue/model/GOSoundingPipe.cpp [hunk 12]
```diff
@@ -486,17 +509,11 @@ void GOSoundingPipe::UpdateAudioGroup() {
...
   else
-    m_TemperamentOffset = temperament.GetOffset(
-      m_PipeConfigNode.GetEffectiveIgnorePitch(),
-      m_MidiKeyNumber,
-      m_SoundProvider.GetMidiKeyNumber(),
-      m_SoundProvider.GetMidiPitchFract(),
-      m_HarmonicNumber,
-      m_PitchCorrection,
-      m_PipeConfigNode.GetDefaultTuning());
+    m_TemperamentOffset = temperament.GetOffset(m_MidiKeyNumber % 12);
```
Comment (oleg68) [inline src/grandorgue/model/GOSoundingPipe.cpp:516]:
```suggestion
    m_TemperamentOffset = m_RetunePipe ? temperament.GetOffset(m_MidiKeyNumber % 12) : 0;
```

Then we can get rid of `if`.

**Review by oleg68 — CHANGES_REQUESTED** [review]

**Comment by larspalo** [comment]
@oleg68 @rousseldenis The pre-commit hooks very often fails with the "corrections" that itself does on the code formatting. It's extremely frustrating. If it doesn't work and help... It simply doesn't have any place! It's only an hindrance when one need to get creative with the formatting just to bypass the so called "errors" the pre-commit has done itself and then "detects" in the formatting, which needless to say has absolutely nothing to do with the actual functionality of the code...

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,3 +1,4 @@
+- Refactored tuning according to discussion in https://github.com/GrandOrgue/grandorgue/issues/1333
```
[the remark this answers is not in the document]
Comment (larspalo) [reply CHANGELOG.md:1]: Agree.

File: src/core/temperaments/GOTemperament.cpp (hunk 1 above)
Comment (larspalo) [reply src/core/temperaments/GOTemperament.cpp:17]: Ok. But it's not better, it's just possible to override at will in the calls to the constructor and not hard coded to either specific case.

File: src/grandorgue/model/GOSoundingPipe.cpp (hunk 12 above)
Comment (larspalo) [reply src/grandorgue/model/GOSoundingPipe.cpp:516]: It's possible. But I won't do it. One thing is the pre-commit hook failing and the other is that ternary operator is inherently more unclear to read than the more verbose if ... else blocks.

File: src/core/temperaments/GOTemperament.h (hunk 5 above)
Comment (larspalo) [reply src/core/temperaments/GOTemperament.h:38]: It is not as clear as GetOffset()

**Review by oleg68 — APPROVED** [review]

**Comment by oleg68** [comment]
> The pre-commit hooks very often fails with the "corrections" that itself does on the code formatting. It's extremely frustrating.

When the pre-commit fails, it corrects the formatting issues itself and changes the files. So you can just stage them and commit to the git repository.

If you do not like the formatting proposed by the hook, you may try to adopt https://github.com/GrandOrgue/grandorgue/blob/master/.clang-format to your wishes (of cause, in a separate PR because usually it requires reformatting the whole sourcecode), but it is not so easy.

**Comment by rousseldenis** [comment]
@larspalo the idea behind pre-commit is to not take care about formatting for developer.

The thing to understand as @oleg68 said is that it can take two passes to get your changes committed.

So, first, do your changes.

Secondly, do a git commit. If your code is well formatted, it's directly committed. if not, pre-commit do the main changes on its own. So, you just need to run the git commit command one more time and that's it.

Don't take care about how pre-commit changed your code even it seems weird to you.

If you think it's wrong, maybe some options need to be changed.

**Comment by larspalo** [comment]
@oleg68 @rousseldenis So if it fails the second time I run it (without me doing any changes) there's a bug, or conflicting instructions it tries to follow?

**Comment by oleg68** [comment]
> @oleg68 @rousseldenis So if it fails the second time I run it (without me doing any changes) there's a bug, or conflicting instructions it tries to follow?

You have to stage changed files before the second attempt of commit.

**Comment by larspalo** [comment]
> You have to stage changed files before the second attempt of commit.

Using git add separately? Because git commit -a -m 'Some comment' fail.

**Comment by rousseldenis** [comment]
> > You have to stage changed files before the second attempt of commit.
> 
> Using git add separately? Because git commit -a -m 'Some comment' fail.

Yes, sorry you have to add them again

**Review by rousseldenis — APPROVED** [review]

**Pull request merged**